File: backend/app/utils/security.py
"""
Security Utilities for Local Producer Web Application

This module provides comprehensive security utilities including input validation,
sanitization, encryption, and Romanian GDPR compliance measures.
"""

import re
import html
import hashlib
import secrets
import logging
from datetime import datetime, timezone, timedelta
from typing import Dict, List, Any, Optional, Union
from werkzeug.security import check_password_hash, generate_password_hash
import bleach
from email_validator import validate_email, EmailNotValidError


# Romanian phone number regex
ROMANIAN_PHONE_REGEX = re.compile(r'^(\+4|0040|0)([0-9]{9})$')

# Romanian postal code regex  
ROMANIAN_POSTAL_CODE_REGEX = re.compile(r'^[0-9]{6}$')

# Strong password requirements
PASSWORD_MIN_LENGTH = 8
PASSWORD_MAX_LENGTH = 128
PASSWORD_REQUIREMENTS = {
    'min_length': PASSWORD_MIN_LENGTH,
    'max_length': PASSWORD_MAX_LENGTH,
    'require_uppercase': True,
    'require_lowercase': True,
    'require_digits': True,
    'require_special': True,
    'forbidden_patterns': [
        'password', 'parola', '123456', 'qwerty', 'admin',
        'test', 'user', 'guest', 'demo'
    ]
}

# Allowed file extensions for uploads
ALLOWED_IMAGE_EXTENSIONS = {'.jpg', '.jpeg', '.png', '.webp', '.gif'}
MAX_FILE_SIZE = 5 * 1024 * 1024  # 5MB

# Security headers configuration
SECURITY_HEADERS = {
    'X-Content-Type-Options': 'nosniff',
    'X-Frame-Options': 'DENY',
    'X-XSS-Protection': '1; mode=block',
    'Strict-Transport-Security': 'max-age=31536000; includeSubDomains',
    'Content-Security-Policy': (
        "default-src 'self'; "
        "script-src 'self' 'unsafe-inline' https://www.googletagmanager.com https://www.google-analytics.com; "
        "style-src 'self' 'unsafe-inline' https://fonts.googleapis.com; "
        "font-src 'self' https://fonts.gstatic.com; "
        "img-src 'self' data: https:; "
        "connect-src 'self' https://www.google-analytics.com; "
        "frame-ancestors 'none'; "
        "base-uri 'self'; "
        "form-action 'self';"
    ),
    'Referrer-Policy': 'strict-origin-when-cross-origin',
    'Permissions-Policy': 'geolocation=(), microphone=(), camera=()'
}


class SecurityValidator:
    """Security validation and sanitization utilities"""
    
    @staticmethod
    def validate_password(password: str) -> Dict[str, Any]:
        """
        Validate password strength according to security requirements.
        
        Args:
            password (str): Password to validate
            
        Returns:
            Dict containing validation result and details
        """
        result = {
            'is_valid': True,
            'errors': [],
            'strength_score': 0,
            'recommendations': []
        }
        
        if not password:
            result['is_valid'] = False
            result['errors'].append("Parola este obligatorie")
            return result
        
        # Length validation
        if len(password) < PASSWORD_REQUIREMENTS['min_length']:
            result['is_valid'] = False
            result['errors'].append(f"Parola trebuie să aibă cel puțin {PASSWORD_REQUIREMENTS['min_length']} caractere")
        
        if len(password) > PASSWORD_REQUIREMENTS['max_length']:
            result['is_valid'] = False
            result['errors'].append(f"Parola nu poate avea mai mult de {PASSWORD_REQUIREMENTS['max_length']} caractere")
        
        # Character requirements
        if PASSWORD_REQUIREMENTS['require_uppercase'] and not re.search(r'[A-Z]', password):
            result['is_valid'] = False
            result['errors'].append("Parola trebuie să conțină cel puțin o literă mare")
            result['recommendations'].append("Adaugă cel puțin o literă mare (A-Z)")
        
        if PASSWORD_REQUIREMENTS['require_lowercase'] and not re.search(r'[a-z]', password):
            result['is_valid'] = False
            result['errors'].append("Parola trebuie să conțină cel puțin o literă mică")
            result['recommendations'].append("Adaugă cel puțin o literă mică (a-z)")
        
        if PASSWORD_REQUIREMENTS['require_digits'] and not re.search(r'[0-9]', password):
            result['is_valid'] = False
            result['errors'].append("Parola trebuie să conțină cel puțin o cifră")
            result['recommendations'].append("Adaugă cel puțin o cifră (0-9)")
        
        if PASSWORD_REQUIREMENTS['require_special'] and not re.search(r'[!@#$%^&*(),.?":{}|<>]', password):
            result['is_valid'] = False
            result['errors'].append("Parola trebuie să conțină cel puțin un caracter special")
            result['recommendations'].append("Adaugă cel puțin un caracter special (!@#$%^&*)")
        
        # Forbidden patterns
        password_lower = password.lower()
        for pattern in PASSWORD_REQUIREMENTS['forbidden_patterns']:
            if pattern in password_lower:
                result['is_valid'] = False
                result['errors'].append(f"Parola nu poate conține '{pattern}'")
        
        # Calculate strength score
        strength_score = 0
        if len(password) >= 8:
            strength_score += 20
        if len(password) >= 12:
            strength_score += 10
        if re.search(r'[A-Z]', password):
            strength_score += 15
        if re.search(r'[a-z]', password):
            strength_score += 15
        if re.search(r'[0-9]', password):
            strength_score += 15
        if re.search(r'[!@#$%^&*(),.?":{}|<>]', password):
            strength_score += 15
        if len(set(password)) > len(password) * 0.7:  # Character diversity
            strength_score += 10
        
        result['strength_score'] = min(strength_score, 100)
        
        return result
    
    @staticmethod
    def validate_email(email: str) -> Dict[str, Any]:
        """
        Validate email address format and security.
        
        Args:
            email (str): Email address to validate
            
        Returns:
            Dict containing validation result
        """
        result = {
            'is_valid': True,
            'errors': [],
            'normalized_email': None
        }
        
        if not email:
            result['is_valid'] = False
            result['errors'].append("Adresa de email este obligatorie")
            return result
        
        try:
            # Use email-validator library for comprehensive validation
            validation = validate_email(email)
            result['normalized_email'] = validation.email
            
            # Additional Romanian domain validation
            domain = validation.domain
            if len(domain) > 253:  # RFC limit
                result['is_valid'] = False
                result['errors'].append("Domeniul email-ului este prea lung")
            
        except EmailNotValidError as e:
            result['is_valid'] = False
            result['errors'].append("Formatul adresei de email nu este valid")
        
        return result
    
    @staticmethod
    def validate_romanian_phone(phone: str) -> Dict[str, Any]:
        """
        Validate Romanian phone number format.
        
        Args:
            phone (str): Phone number to validate
            
        Returns:
            Dict containing validation result
        """
        result = {
            'is_valid': True,
            'errors': [],
            'normalized_phone': None
        }
        
        if not phone:
            result['is_valid'] = False
            result['errors'].append("Numărul de telefon este obligatoriu")
            return result
        
        # Remove spaces and dashes
        clean_phone = re.sub(r'[\s\-\(\)]', '', phone)
        
        if not ROMANIAN_PHONE_REGEX.match(clean_phone):
            result['is_valid'] = False
            result['errors'].append("Formatul numărului de telefon nu este valid pentru România")
            return result
        
        # Normalize to +40 format
        if clean_phone.startswith('0040'):
            result['normalized_phone'] = '+4' + clean_phone[4:]
        elif clean_phone.startswith('+4'):
            result['normalized_phone'] = clean_phone
        elif clean_phone.startswith('0'):
            result['normalized_phone'] = '+4' + clean_phone[1:]
        else:
            result['normalized_phone'] = '+4' + clean_phone
        
        return result
    
    @staticmethod
    def sanitize_input(input_text: str, max_length: int = 1000) -> str:
        """
        Sanitize user input to prevent XSS and injection attacks.
        
        Args:
            input_text (str): Input text to sanitize
            max_length (int): Maximum allowed length
            
        Returns:
            Sanitized text
        """
        if not input_text:
            return ""
        
        # Truncate if too long
        if len(input_text) > max_length:
            input_text = input_text[:max_length]
        
        # HTML escape
        sanitized = html.escape(input_text)
        
        # Use bleach for additional sanitization
        allowed_tags = []  # No HTML tags allowed in general input
        sanitized = bleach.clean(sanitized, tags=allowed_tags, strip=True)
        
        # Remove null bytes and control characters
        sanitized = ''.join(char for char in sanitized if ord(char) >= 32 or char in '\n\r\t')
        
        return sanitized.strip()
    
    @staticmethod
    def validate_product_name(name: str) -> Dict[str, Any]:
        """
        Validate product name with Romanian context.
        
        Args:
            name (str): Product name to validate
            
        Returns:
            Dict containing validation result
        """
        result = {
            'is_valid': True,
            'errors': [],
            'sanitized_name': None
        }
        
        if not name:
            result['is_valid'] = False
            result['errors'].append("Numele produsului este obligatoriu")
            return result
        
        # Sanitize
        sanitized_name = SecurityValidator.sanitize_input(name, max_length=100)
        
        if len(sanitized_name) < 2:
            result['is_valid'] = False
            result['errors'].append("Numele produsului trebuie să aibă cel puțin 2 caractere")
        
        if len(sanitized_name) > 100:
            result['is_valid'] = False
            result['errors'].append("Numele produsului nu poate depăși 100 de caractere")
        
        # Check for inappropriate content (basic Romanian profanity filter)
        forbidden_words = ['test', 'spam', 'fake']  # Add Romanian inappropriate words
        name_lower = sanitized_name.lower()
        for word in forbidden_words:
            if word in name_lower:
                result['is_valid'] = False
                result['errors'].append("Numele produsului conține conținut neadecvat")
        
        result['sanitized_name'] = sanitized_name
        return result
    
    @staticmethod
    def validate_price(price: Union[str, float, int]) -> Dict[str, Any]:
        """
        Validate product price.
        
        Args:
            price: Price to validate
            
        Returns:
            Dict containing validation result
        """
        result = {
            'is_valid': True,
            'errors': [],
            'normalized_price': None
        }
        
        try:
            price_float = float(price)
            
            if price_float < 0:
                result['is_valid'] = False
                result['errors'].append("Prețul nu poate fi negativ")
            elif price_float == 0:
                result['is_valid'] = False
                result['errors'].append("Prețul trebuie să fie mai mare decât 0")
            elif price_float > 99999.99:
                result['is_valid'] = False
                result['errors'].append("Prețul nu poate depăși 99,999.99 RON")
            else:
                # Round to 2 decimal places
                result['normalized_price'] = round(price_float, 2)
                
        except (ValueError, TypeError):
            result['is_valid'] = False
            result['errors'].append("Prețul trebuie să fie un număr valid")
        
        return result
    
    @staticmethod
    def validate_file_upload(filename: str, file_size: int) -> Dict[str, Any]:
        """
        Validate file upload security.
        
        Args:
            filename (str): Name of uploaded file
            file_size (int): Size of uploaded file in bytes
            
        Returns:
            Dict containing validation result
        """
        result = {
            'is_valid': True,
            'errors': [],
            'safe_filename': None
        }
        
        if not filename:
            result['is_valid'] = False
            result['errors'].append("Numele fișierului este obligatoriu")
            return result
        
        # Check file size
        if file_size > MAX_FILE_SIZE:
            result['is_valid'] = False
            result['errors'].append(f"Fișierul este prea mare. Mărimea maximă permisă este {MAX_FILE_SIZE // (1024*1024)}MB")
        
        # Check extension
        file_ext = '.' + filename.lower().split('.')[-1] if '.' in filename else ''
        if file_ext not in ALLOWED_IMAGE_EXTENSIONS:
            result['is_valid'] = False
            result['errors'].append(f"Tipul de fișier nu este permis. Extensii permise: {', '.join(ALLOWED_IMAGE_EXTENSIONS)}")
        
        # Generate safe filename
        safe_name = SecurityValidator.sanitize_input(filename, max_length=255)
        safe_name = re.sub(r'[^a-zA-Z0-9._-]', '_', safe_name)
        
        # Prevent directory traversal
        safe_name = safe_name.replace('..', '_').replace('/', '_').replace('\\', '_')
        
        result['safe_filename'] = safe_name
        return result


class SecurityLogger:
    """Security event logging"""
    
    @staticmethod
    def log_security_event(event_type: str, details: Dict[str, Any], 
                          severity: str = 'INFO', user_id: str = None,
                          ip_address: str = None):
        """
        Log security events for monitoring and compliance.
        
        Args:
            event_type (str): Type of security event
            details (Dict): Event details
            severity (str): Event severity (INFO, WARNING, ERROR, CRITICAL)
            user_id (str): User ID if applicable
            ip_address (str): Client IP address
        """
        log_entry = {
            'timestamp': datetime.now(timezone.utc).isoformat(),
            'event_type': event_type,
            'severity': severity,
            'details': details,
            'user_id': user_id,
            'ip_address': ip_address,
            'application': 'local_producer_app'
        }
        
        # Log based on severity
        logger = logging.getLogger('security')
        if severity == 'CRITICAL':
            logger.critical(f"SECURITY EVENT: {event_type}", extra=log_entry)
        elif severity == 'ERROR':
            logger.error(f"SECURITY EVENT: {event_type}", extra=log_entry)
        elif severity == 'WARNING':
            logger.warning(f"SECURITY EVENT: {event_type}", extra=log_entry)
        else:
            logger.info(f"SECURITY EVENT: {event_type}", extra=log_entry)
    
    @staticmethod
    def log_authentication_event(event_type: str, phone_number: str = None,
                                success: bool = True, ip_address: str = None,
                                details: Dict[str, Any] = None):
        """Log authentication events"""
        SecurityLogger.log_security_event(
            event_type=f"authentication_{event_type}",
            details={
                'phone_number': phone_number[-4:] if phone_number else None,  # Log only last 4 digits
                'success': success,
                'details': details or {}
            },
            severity='WARNING' if not success else 'INFO',
            ip_address=ip_address
        )
    
    @staticmethod
    def log_data_access(data_type: str, operation: str, user_id: str = None,
                       ip_address: str = None, details: Dict[str, Any] = None):
        """Log data access events for GDPR compliance"""
        SecurityLogger.log_security_event(
            event_type='data_access',
            details={
                'data_type': data_type,
                'operation': operation,
                'details': details or {}
            },
            severity='INFO',
            user_id=user_id,
            ip_address=ip_address
        )


class RateLimiter:
    """Rate limiting for API endpoints"""
    
    def __init__(self):
        self.requests = {}  # In-memory storage (use Redis in production)
        self.cleanup_interval = 3600  # 1 hour
        self.last_cleanup = datetime.now(timezone.utc)
    
    def is_allowed(self, identifier: str, limit: int, window: int) -> Dict[str, Any]:
        """
        Check if request is allowed based on rate limit.
        
        Args:
            identifier (str): Unique identifier (IP, user ID, etc.)
            limit (int): Maximum requests allowed
            window (int): Time window in seconds
            
        Returns:
            Dict with allowed status and metadata
        """
        now = datetime.now(timezone.utc)
        
        # Cleanup old entries periodically
        if (now - self.last_cleanup).seconds > self.cleanup_interval:
            self._cleanup_old_entries()
            self.last_cleanup = now
        
        # Get request history for identifier
        if identifier not in self.requests:
            self.requests[identifier] = []
        
        request_times = self.requests[identifier]
        
        # Remove requests outside the window
        cutoff_time = now - timedelta(seconds=window)
        request_times = [t for t in request_times if t > cutoff_time]
        self.requests[identifier] = request_times
        
        # Check if limit exceeded
        if len(request_times) >= limit:
            return {
                'allowed': False,
                'requests_made': len(request_times),
                'limit': limit,
                'window': window,
                'retry_after': window
            }
        
        # Record this request
        request_times.append(now)
        
        return {
            'allowed': True,
            'requests_made': len(request_times),
            'limit': limit,
            'window': window,
            'remaining': limit - len(request_times)
        }
    
    def _cleanup_old_entries(self):
        """Remove old rate limit entries"""
        cutoff_time = datetime.now(timezone.utc) - timedelta(hours=24)
        
        for identifier in list(self.requests.keys()):
            self.requests[identifier] = [
                t for t in self.requests[identifier] if t > cutoff_time
            ]
            
            # Remove empty entries
            if not self.requests[identifier]:
                del self.requests[identifier]


# Global rate limiter instance
rate_limiter = RateLimiter()


def generate_secure_token(length: int = 32) -> str:
    """Generate cryptographically secure token"""
    return secrets.token_urlsafe(length)


def hash_sensitive_data(data: str) -> str:
    """Hash sensitive data for storage"""
    return hashlib.sha256(data.encode('utf-8')).hexdigest()


def validate_csrf_token(token: str, session_token: str) -> bool:
    """Validate CSRF token"""
    return secrets.compare_digest(token, session_token)


def apply_security_headers(response):
    """Apply security headers to response"""
    for header, value in SECURITY_HEADERS.items():
        response.headers[header] = value
    return response

File: frontend/src/utils/security.js
/**
 * Frontend Security Utilities for Local Producer Web Application
 * 
 * This module provides comprehensive frontend security utilities including
 * input validation, sanitization, and security-focused user interactions.
 */

// Romanian phone number validation regex
const ROMANIAN_PHONE_REGEX = /^(\+4|0040|0)([0-9]{9})$/;

// Romanian postal code validation regex
const ROMANIAN_POSTAL_CODE_REGEX = /^[0-9]{6}$/;

// Strong password requirements
export const PASSWORD_REQUIREMENTS = {
  minLength: 8,
  maxLength: 128,
  requireUppercase: true,
  requireLowercase: true,
  requireDigits: true,
  requireSpecial: true,
  forbiddenPatterns: [
    'password', 'parola', '123456', 'qwerty', 'admin',
    'test', 'user', 'guest', 'demo'
  ]
};

// Allowed file extensions for uploads
export const ALLOWED_IMAGE_EXTENSIONS = ['.jpg', '.jpeg', '.png', '.webp', '.gif'];
export const MAX_FILE_SIZE = 5 * 1024 * 1024; // 5MB

/**
 * Frontend Security Validator Class
 */
export class SecurityValidator {
  /**
   * Validate password strength according to security requirements
   * @param {string} password - Password to validate
   * @returns {Object} Validation result with details
   */
  static validatePassword(password) {
    const result = {
      isValid: true,
      errors: [],
      strengthScore: 0,
      recommendations: []
    };

    if (!password) {
      result.isValid = false;
      result.errors.push("Parola este obligatorie");
      return result;
    }

    // Length validation
    if (password.length < PASSWORD_REQUIREMENTS.minLength) {
      result.isValid = false;
      result.errors.push(`Parola trebuie să aibă cel puțin ${PASSWORD_REQUIREMENTS.minLength} caractere`);
    }

    if (password.length > PASSWORD_REQUIREMENTS.maxLength) {
      result.isValid = false;
      result.errors.push(`Parola nu poate avea mai mult de ${PASSWORD_REQUIREMENTS.maxLength} caractere`);
    }

    // Character requirements
    if (PASSWORD_REQUIREMENTS.requireUppercase && !/[A-Z]/.test(password)) {
      result.isValid = false;
      result.errors.push("Parola trebuie să conțină cel puțin o literă mare");
      result.recommendations.push("Adaugă cel puțin o literă mare (A-Z)");
    }

    if (PASSWORD_REQUIREMENTS.requireLowercase && !/[a-z]/.test(password)) {
      result.isValid = false;
      result.errors.push("Parola trebuie să conțină cel puțin o literă mică");
      result.recommendations.push("Adaugă cel puțin o literă mică (a-z)");
    }

    if (PASSWORD_REQUIREMENTS.requireDigits && !/[0-9]/.test(password)) {
      result.isValid = false;
      result.errors.push("Parola trebuie să conțină cel puțin o cifră");
      result.recommendations.push("Adaugă cel puțin o cifră (0-9)");
    }

    if (PASSWORD_REQUIREMENTS.requireSpecial && !/[!@#$%^&*(),.?":{}|<>]/.test(password)) {
      result.isValid = false;
      result.errors.push("Parola trebuie să conțină cel puțin un caracter special");
      result.recommendations.push("Adaugă cel puțin un caracter special (!@#$%^&*)");
    }

    // Forbidden patterns
    const passwordLower = password.toLowerCase();
    for (const pattern of PASSWORD_REQUIREMENTS.forbiddenPatterns) {
      if (passwordLower.includes(pattern)) {
        result.isValid = false;
        result.errors.push(`Parola nu poate conține '${pattern}'`);
      }
    }

    // Calculate strength score
    let strengthScore = 0;
    if (password.length >= 8) strengthScore += 20;
    if (password.length >= 12) strengthScore += 10;
    if (/[A-Z]/.test(password)) strengthScore += 15;
    if (/[a-z]/.test(password)) strengthScore += 15;
    if (/[0-9]/.test(password)) strengthScore += 15;
    if (/[!@#$%^&*(),.?":{}|<>]/.test(password)) strengthScore += 15;
    if (new Set(password).size > password.length * 0.7) strengthScore += 10; // Character diversity

    result.strengthScore = Math.min(strengthScore, 100);

    return result;
  }

  /**
   * Validate email address format
   * @param {string} email - Email address to validate
   * @returns {Object} Validation result
   */
  static validateEmail(email) {
    const result = {
      isValid: true,
      errors: [],
      normalizedEmail: null
    };

    if (!email) {
      result.isValid = false;
      result.errors.push("Adresa de email este obligatorie");
      return result;
    }

    // Basic email format validation
    const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
    if (!emailRegex.test(email)) {
      result.isValid = false;
      result.errors.push("Formatul adresei de email nu este valid");
      return result;
    }

    // Additional validation
    const [localPart, domain] = email.split('@');
    
    if (localPart.length > 64) {
      result.isValid = false;
      result.errors.push("Partea locală a email-ului este prea lungă");
    }

    if (domain.length > 253) {
      result.isValid = false;
      result.errors.push("Domeniul email-ului este prea lung");
    }

    result.normalizedEmail = email.toLowerCase().trim();
    return result;
  }

  /**
   * Validate Romanian phone number format
   * @param {string} phone - Phone number to validate
   * @returns {Object} Validation result
   */
  static validateRomanianPhone(phone) {
    const result = {
      isValid: true,
      errors: [],
      normalizedPhone: null
    };

    if (!phone) {
      result.isValid = false;
      result.errors.push("Numărul de telefon este obligatoriu");
      return result;
    }

    // Remove spaces and dashes
    const cleanPhone = phone.replace(/[\s\-\(\)]/g, '');

    if (!ROMANIAN_PHONE_REGEX.test(cleanPhone)) {
      result.isValid = false;
      result.errors.push("Formatul numărului de telefon nu este valid pentru România");
      return result;
    }

    // Normalize to +40 format
    if (cleanPhone.startsWith('0040')) {
      result.normalizedPhone = '+4' + cleanPhone.slice(4);
    } else if (cleanPhone.startsWith('+4')) {
      result.normalizedPhone = cleanPhone;
    } else if (cleanPhone.startsWith('0')) {
      result.normalizedPhone = '+4' + cleanPhone.slice(1);
    } else {
      result.normalizedPhone = '+4' + cleanPhone;
    }

    return result;
  }

  /**
   * Sanitize user input to prevent XSS attacks
   * @param {string} input - Input text to sanitize
   * @param {number} maxLength - Maximum allowed length
   * @returns {string} Sanitized text
   */
  static sanitizeInput(input, maxLength = 1000) {
    if (!input) return '';

    // Truncate if too long
    if (input.length > maxLength) {
      input = input.slice(0, maxLength);
    }

    // HTML escape
    const htmlEscapeMap = {
      '&': '&amp;',
      '<': '&lt;',
      '>': '&gt;',
      '"': '&quot;',
      "'": '&#x27;',
      '/': '&#x2F;'
    };

    let sanitized = input.replace(/[&<>"'/]/g, (char) => htmlEscapeMap[char]);

    // Remove null bytes and control characters
    sanitized = sanitized.replace(/[\x00-\x1f\x7f-\x9f]/g, '');

    return sanitized.trim();
  }

  /**
   * Validate product name with Romanian context
   * @param {string} name - Product name to validate
   * @returns {Object} Validation result
   */
  static validateProductName(name) {
    const result = {
      isValid: true,
      errors: [],
      sanitizedName: null
    };

    if (!name) {
      result.isValid = false;
      result.errors.push("Numele produsului este obligatoriu");
      return result;
    }

    // Sanitize
    const sanitizedName = this.sanitizeInput(name, 100);

    if (sanitizedName.length < 2) {
      result.isValid = false;
      result.errors.push("Numele produsului trebuie să aibă cel puțin 2 caractere");
    }

    if (sanitizedName.length > 100) {
      result.isValid = false;
      result.errors.push("Numele produsului nu poate depăși 100 de caractere");
    }

    // Check for inappropriate content (basic filter)
    const forbiddenWords = ['test', 'spam', 'fake'];
    const nameLower = sanitizedName.toLowerCase();
    for (const word of forbiddenWords) {
      if (nameLower.includes(word)) {
        result.isValid = false;
        result.errors.push("Numele produsului conține conținut neadecvat");
        break;
      }
    }

    result.sanitizedName = sanitizedName;
    return result;
  }

  /**
   * Validate product price
   * @param {string|number} price - Price to validate
   * @returns {Object} Validation result
   */
  static validatePrice(price) {
    const result = {
      isValid: true,
      errors: [],
      normalizedPrice: null
    };

    try {
      const priceFloat = parseFloat(price);

      if (isNaN(priceFloat)) {
        result.isValid = false;
        result.errors.push("Prețul trebuie să fie un număr valid");
      } else if (priceFloat < 0) {
        result.isValid = false;
        result.errors.push("Prețul nu poate fi negativ");
      } else if (priceFloat === 0) {
        result.isValid = false;
        result.errors.push("Prețul trebuie să fie mai mare decât 0");
      } else if (priceFloat > 99999.99) {
        result.isValid = false;
        result.errors.push("Prețul nu poate depăși 99,999.99 RON");
      } else {
        // Round to 2 decimal places
        result.normalizedPrice = Math.round(priceFloat * 100) / 100;
      }
    } catch (error) {
      result.isValid = false;
      result.errors.push("Prețul trebuie să fie un număr valid");
    }

    return result;
  }

  /**
   * Validate file upload security
   * @param {File} file - File to validate
   * @returns {Object} Validation result
   */
  static validateFileUpload(file) {
    const result = {
      isValid: true,
      errors: [],
      safeFilename: null
    };

    if (!file) {
      result.isValid = false;
      result.errors.push("Fișierul este obligatoriu");
      return result;
    }

    // Check file size
    if (file.size > MAX_FILE_SIZE) {
      result.isValid = false;
      result.errors.push(`Fișierul este prea mare. Mărimea maximă permisă este ${MAX_FILE_SIZE / (1024 * 1024)}MB`);
    }

    // Check extension
    const fileName = file.name.toLowerCase();
    const fileExt = '.' + fileName.split('.').pop();
    if (!ALLOWED_IMAGE_EXTENSIONS.includes(fileExt)) {
      result.isValid = false;
      result.errors.push(`Tipul de fișier nu este permis. Extensii permise: ${ALLOWED_IMAGE_EXTENSIONS.join(', ')}`);
    }

    // Generate safe filename
    const safeName = this.sanitizeInput(file.name, 255);
    const sanitizedName = safeName.replace(/[^a-zA-Z0-9._-]/g, '_');
    
    // Prevent directory traversal
    const finalName = sanitizedName.replace(/\.\./g, '_').replace(/[\/\\]/g, '_');

    result.safeFilename = finalName;
    return result;
  }
}

/**
 * Client-side CSRF token management
 */
export class CSRFProtection {
  static getMetaToken() {
    const metaTag = document.querySelector('meta[name="csrf-token"]');
    return metaTag ? metaTag.getAttribute('content') : null;
  }

  static setRequestHeaders(headers = {}) {
    const token = this.getMetaToken();
    if (token) {
      headers['X-CSRF-Token'] = token;
    }
    return headers;
  }

  static createSecureForm(formElement) {
    const token = this.getMetaToken();
    if (token && formElement) {
      const csrfInput = document.createElement('input');
      csrfInput.type = 'hidden';
      csrfInput.name = 'csrf_token';
      csrfInput.value = token;
      formElement.appendChild(csrfInput);
    }
  }
}

/**
 * Secure data masking for display
 */
export class DataMasking {
  static maskPhoneNumber(phoneNumber) {
    if (!phoneNumber) return '';

    // Remove country code and formatting
    const cleanPhone = phoneNumber.replace(/^\+40/, '').replace(/[\s\-]/g, '');
    
    if (cleanPhone.length >= 4) {
      return `+40***${cleanPhone.slice(-4)}`;
    } else {
      return '+40***';
    }
  }

  static maskEmail(email) {
    if (!email || !email.includes('@')) return '***';

    const [local, domain] = email.split('@');
    
    if (local.length <= 2) {
      return '*'.repeat(local.length) + '@' + domain;
    } else {
      return local[0] + '*'.repeat(local.length - 2) + local.slice(-1) + '@' + domain;
    }
  }

  static maskAddress(address) {
    if (!address) return '';

    const words = address.split(' ');
    if (words.length <= 2) return '***';

    // Show first and last word, mask middle
    return `${words[0]} *** ${words[words.length - 1]}`;
  }
}

/**
 * Frontend rate limiting (client-side protection)
 */
export class ClientRateLimit {
  constructor() {
    this.requests = new Map();
    this.cleanupInterval = 60000; // 1 minute
    this.lastCleanup = Date.now();
  }

  isAllowed(identifier, limit, windowMs) {
    const now = Date.now();

    // Cleanup old entries periodically
    if (now - this.lastCleanup > this.cleanupInterval) {
      this.cleanup();
      this.lastCleanup = now;
    }

    // Get request history for identifier
    if (!this.requests.has(identifier)) {
      this.requests.set(identifier, []);
    }

    const requestTimes = this.requests.get(identifier);

    // Remove requests outside the window
    const cutoffTime = now - windowMs;
    const validRequests = requestTimes.filter(time => time > cutoffTime);
    this.requests.set(identifier, validRequests);

    // Check if limit exceeded
    if (validRequests.length >= limit) {
      return {
        allowed: false,
        requestsMade: validRequests.length,
        limit,
        retryAfter: windowMs
      };
    }

    // Record this request
    validRequests.push(now);

    return {
      allowed: true,
      requestsMade: validRequests.length,
      limit,
      remaining: limit - validRequests.length
    };
  }

  cleanup() {
    const cutoffTime = Date.now() - (24 * 60 * 60 * 1000); // 24 hours
    
    for (const [identifier, requestTimes] of this.requests.entries()) {
      const validRequests = requestTimes.filter(time => time > cutoffTime);
      if (validRequests.length === 0) {
        this.requests.delete(identifier);
      } else {
        this.requests.set(identifier, validRequests);
      }
    }
  }
}

/**
 * Secure local storage management
 */
export class SecureStorage {
  static setItem(key, value, encrypt = false) {
    try {
      const data = {
        value,
        timestamp: Date.now(),
        encrypted: encrypt
      };

      if (encrypt) {
        // Simple base64 encoding (not real encryption, just obfuscation)
        data.value = btoa(JSON.stringify(value));
      }

      localStorage.setItem(key, JSON.stringify(data));
      return true;
    } catch (error) {
      console.error('Failed to store data:', error);
      return false;
    }
  }

  static getItem(key, maxAge = null) {
    try {
      const stored = localStorage.getItem(key);
      if (!stored) return null;

      const data = JSON.parse(stored);

      // Check expiration
      if (maxAge && Date.now() - data.timestamp > maxAge) {
        localStorage.removeItem(key);
        return null;
      }

      if (data.encrypted) {
        // Simple base64 decoding
        return JSON.parse(atob(data.value));
      }

      return data.value;
    } catch (error) {
      console.error('Failed to retrieve data:', error);
      localStorage.removeItem(key);
      return null;
    }
  }

  static removeItem(key) {
    localStorage.removeItem(key);
  }

  static clear() {
    localStorage.clear();
  }
}

/**
 * Content Security Policy helper
 */
export class CSPHelper {
  static reportViolation(violationReport) {
    // Log CSP violations for security monitoring
    console.warn('CSP Violation:', violationReport);
    
    // In production, send to security monitoring endpoint
    if (process.env.NODE_ENV === 'production') {
      fetch('/api/security/csp-violation', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(violationReport)
      }).catch(console.error);
    }
  }

  static setupViolationReporting() {
    document.addEventListener('securitypolicyviolation', (event) => {
      this.reportViolation({
        documentURI: event.documentURI,
        referrer: event.referrer,
        blockedURI: event.blockedURI,
        violatedDirective: event.violatedDirective,
        effectiveDirective: event.effectiveDirective,
        originalPolicy: event.originalPolicy,
        disposition: event.disposition,
        statusCode: event.statusCode
      });
    });
  }
}

// Initialize CSP violation reporting
if (typeof document !== 'undefined') {
  CSPHelper.setupViolationReporting();
}

// Global rate limiter instance
export const clientRateLimit = new ClientRateLimit();

// Default export with all utilities
export default {
  SecurityValidator,
  CSRFProtection,
  DataMasking,
  ClientRateLimit,
  SecureStorage,
  CSPHelper,
  clientRateLimit,
  PASSWORD_REQUIREMENTS,
  ALLOWED_IMAGE_EXTENSIONS,
  MAX_FILE_SIZE
};